javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,453 @@
+      head = oldHead.next
+      head.prev = null
+
+      if (_size == 2) {
+        head.next = null
+        last = Node[Box[E]]()
+      }
+    }
+
+    if (ret eq null) null.asInstanceOf[E]
+    else {
+      _size -= 1
+      ret.inner
+    }
+  }
+
+  def removeLast(): E = {
+    val ret = last.value
+
+    if (_size < 2) {
+      removeFirst()
+    } else {
+      val oldLast = last
+      last = oldLast.prev
+      last.next = null
+
+      if (size == 2) {
+        head = last
+        last = Node[Box[E]]()
+      }
+    }
+
+    if (ret eq null) null.asInstanceOf[E]
+    else {
+      _size -= 1
+      ret.inner
+    }
+  }
+
+  def addFirst(e: E): Unit = {
+    val boxed = Box(e)
+
+    val oldHead =
+      if (head.value ne null) head
+      else null
+
+    head = Node(boxed, next = oldHead)
+
+    _size += 1
+
+    if (oldHead ne null) {
+      oldHead.prev = head
+
+      if ((last.value eq null) && (oldHead.next eq null))
+        last = oldHead
+    }
+  }
+
+  def addLast(e: E): Unit = {
+    val boxed = Box(e)
+
+    if (head.value eq null) addFirst(e)
+    else {
+      val oldLast =
+        if (last.value ne null) last
+        else if (head.value ne null) head
+        else null
+
+      last = Node(boxed, prev = oldLast)
+
+      _size += 1
+
+      if (oldLast ne null)
+        oldLast.next = last
+    }
+  }
+
+  override def contains(o: Any): Boolean = {
+    val boxed = Box(o.asInstanceOf[E])
+    var found = false
+    val iter = iterator()
+    while (iter.hasNext() && !found) {
+      val next = iter.next()
+      if (Box(next) === boxed) {
+        found = true
+      }
+    }
+
+    found
+  }
+
+  override def size(): Int =
+    _size
+
+  override def add(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  override def remove(o: Any): Boolean =
+    _removeOccurrence(listIterator, o)
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator
+    val changed = iter.hasNext()
+    while (iter.hasNext())
+      addLast(iter.next())
+
+    changed
+  }
+
+  override def clear(): Unit = {
+    head = Node()
+    last = Node()
+    _size = 0
+  }
+
+  private def getNodeAt(index: Int): Node[Box[E]] = {
+    if (index == 0) head
+    else if (index == (size - 1)) last
+    else {
+      var current: Node[Box[E]] = null
+      if (index <= (size/2)) {
+        current = head
+        for (_ <- 0 until index) current = current.next
+      } else {
+        current = last
+        for (_ <- index until (size - 1)) current = current.prev
+      }
+      current
+    }
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+    getNodeAt(index).value.inner
+  }
+
+  override def set(index: Int, element: E): E = {
+    checkIndexInBounds(index)
+    val node = getNodeAt(index)
+    val oldValue = node.value.inner
+    node.value = Box(element)
+    oldValue
+  }
+
+  private def addNode(next: Node[Box[E]], e: E): Unit = {
+    if (next.uuid == head.uuid) addFirst(e)
```

Comment:
And if `Node` is not a case class, it's even simply `next == head` (which is then optimized as `next eq head`).